### PR TITLE
ByteString.toArrayUnsafe method for zero copy transformation of bytestrings

### PR DESCRIPTION
**Pull request by johanandren**
Refs #30257

### CODE REVIEW

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
I think this is good, small step, but looking at jfr dump there are also rather many rope ByteStrings

Screenshot 2021-05-26 at 15 59 48

**Comment by johanandren** [comment]
@patriknw did you see the other PR as well? #30263

File: akka-actor/src/main/scala-2.12/akka/util/ByteString.scala [hunk 1]
```diff
@@ -787,6 +794,25 @@ sealed abstract class ByteString extends IndexedSeq[Byte] with IndexedSeqOptimiz
...
+   * you need, and that you understand the implications documented below.
+   *
+   * If the ByteString is backed by a single array it is returned without any copy. If it is backed by a rope
+   * of multiple ByteString instances a new array will be allocated and the contents will be copied
+   * into it before returning it.
+   *
+   * This method of exposing the bytes of a ByteString can save one array
+   * copy and allocation in the happy path scenario and which can lead to better performance,
+   * however it also means that one MUST NOT modify the returned in array, or unexpected
+   * immutable data structure contract-breaking behavior will manifest itself.
```
Comment (raboof) [inline akka-actor/src/main/scala-2.12/akka/util/ByteString.scala:808]:
```suggestion
   * however it also means that one MUST NOT modify the returned array, or unexpected
   * immutable data structure contract-breaking behavior will manifest itself.
```

File: akka-actor/src/main/scala-3/akka/util/ByteString.scala [hunk 2]
```diff
@@ -842,6 +848,25 @@ sealed abstract class ByteString
...
+   * This method of exposing the bytes of a ByteString can save one array
+   * copy and allocation in the happy path scenario and which can lead to better performance,
+   * however it also means that one MUST NOT modify the returned in array, or unexpected
+   * immutable data structure contract-breaking behavior will manifest itself.
+   *
+   * This API is intended for users who need to pass the byte array to some other API, which will
+   * only read the bytes and never mutate then. For all other intents and purposes, please use the usual
+   * toArray method - which provide the immutability guarantees by copying the backing array.
+   */
+  def toArrayUnsafe(): Array[Byte] = toArray
```
Comment (raboof) [inline akka-actor/src/main/scala-3/akka/util/ByteString.scala:868]: No optimized implementation for scala3?

**Review by raboof — APPROVED** [review]

**Pull request merged**